### ISSUE TITLE
Hyperopt: Pickling Error - Could not pickle the task to send it to the workers

### ISSUE THREAD

**Issue by coder194** [freqtrade/freqtrade#1876]
* Python Version: Python 3.6.8
  * CCXT version: ccxt==1.18.523
  * Branch: Master 
  * Last Commit ID: 46b347b6617d96fbc596ecf768f3859cbf26a9b6 
  
## Step 3: Describe the problem:
I am running freqtrade in a Digital Ocean VPS droplet on Ubuntu 16.04.6. When I allocate only 1 CPU to the virtual machine, I am able to successfully run the optimization hyperopt without any errors. However, when I allocate more than 1 CPU (tested with both 2 and 6 CPUs), I get a Pickling Error: "Could not pickle the task to send it to the workers". 

### Steps to reproduce:
  1. Activate virtual environment: 
source .env/bin/activate

  2. Run Optimization Code: 
python3 ./freqtrade/main.py -c config.json hyperopt --customhyperopt DefaultHyperOpts -e 10

  
### Observed Results:

  * _pickle.PicklingError: Could not pickle the task to send it to the workers.
  * I expected all 6 CPUs to be used in optimizing the strategy, however, the error log below is printed out.

### Relevant code exceptions or logs:

  ```
(.env) root@ubuntu:~/freqtrade# python3 ./freqtrade/main.py -c config.json hyperopt --customhyperopt DefaultHyperOpts -e 10
/root/freqtrade/.env/lib/python3.6/site-packages/sklearn/externals/joblib/__init__.py:15: DeprecationWarning: sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. Please import this functionality directly from joblib, which can be installed with: pip install joblib. If this warning is raised when loading pickled models, you may need to re-serialize those models with scikit-learn 0.21+.
  warnings.warn(msg, category=DeprecationWarning)
2019-05-25 04:15:11,178 - freqtrade.configuration - INFO - Verbosity set to 0
2019-05-25 04:15:11,179 - freqtrade.configuration - INFO - Dry run is enabled
2019-05-25 04:15:11,179 - freqtrade.configuration - INFO - Using DB: "sqlite://"
2019-05-25 04:15:11,179 - freqtrade.configuration - INFO - Using max_open_trades: inf ...
2019-05-25 04:15:11,180 - freqtrade.configuration - INFO - Using data folder: user_data/data/binance ...
2019-05-25 04:15:11,180 - freqtrade.configuration - INFO - Using Hyperopt file DefaultHyperOpts
2019-05-25 04:15:11,180 - freqtrade.configuration - INFO - Parameter --epochs detected ... Will run Hyperopt with for 10 epochs ...
2019-05-25 04:15:11,180 - freqtrade.configuration - INFO - Parameter -s/--spaces detected: all
2019-05-25 04:15:11,180 - freqtrade.configuration - INFO - Parameter -j/--job-workers detected: -1
2019-05-25 04:15:11,180 - freqtrade.configuration - INFO - Parameter --min-trades detected: 1
2019-05-25 04:15:11,180 - freqtrade.exchange.exchange - INFO - Instance is running with dry_run enabled
2019-05-25 04:15:11,180 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'enableRateLimit': True}
2019-05-25 04:15:11,183 - freqtrade.exchange.exchange - INFO - Applying additional ccxt config: {'enableRateLimit': True, 'rateLimit': 500}
2019-05-25 04:15:11,194 - freqtrade.exchange.exchange - INFO - Using Exchange "Binance"
2019-05-25 04:15:11,627 - freqtrade.resolvers.exchange_resolver - INFO - Using resolved exchange Binance
2019-05-25 04:15:11,631 - freqtrade.resolvers.strategy_resolver - INFO - Using resolved strategy DefaultStrategy from '/root/freqtrade/freqtrade/strategy'
2019-05-25 04:15:11,631 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'ticker_interval' with value in config file: 1m.
2019-05-25 04:15:11,631 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'trailing_stop' with value in config file: False.
2019-05-25 04:15:11,631 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'stake_currency' with value in config file: BTC.
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'stake_amount' with value in config file: 0.0005.
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'use_sell_signal' with value in config file: False.
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'sell_profit_only' with value in config file: False.
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Override strategy 'ignore_roi_if_buy_signal' with value in config file: False.
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using minimal_roi: {'40': 0.0, '30': 0.01, '20': 0.02, '0': 0.04}
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using ticker_interval: 1m
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stoploss: -0.05
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_stop: False
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_stop_positive_offset: 0.0
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using trailing_only_offset_is_reached: False
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using process_only_new_candles: False
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using order_types: {'buy': 'limit', 'sell': 'limit', 'stoploss': 'limit', 'stoploss_on_exchange': False}
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using order_time_in_force: {'buy': 'gtc', 'sell': 'gtc'}
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stake_currency: BTC
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using stake_amount: 0.0005
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using use_sell_signal: False
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using sell_profit_only: False
2019-05-25 04:15:11,632 - freqtrade.resolvers.strategy_resolver - INFO - Strategy using ignore_roi_if_buy_signal: False
2019-05-25 04:15:11,635 - freqtrade.resolvers.hyperopt_resolver - INFO - Using resolved hyperopt DefaultHyperOpts from '/root/freqtrade/freqtrade/optimize'
2019-05-25 04:15:15,610 - freqtrade.optimize.hyperopt - INFO - Found 6 CPU cores. Let's make them scream!
joblib.externals.loky.process_executor._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/loky/backend/queues.py", line 150, in _feed
    obj_ = dumps(obj, reducers=reducers)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/loky/backend/reduction.py", line 243, in dumps
    dump(obj, buf, reducers=reducers, protocol=protocol)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/loky/backend/reduction.py", line 236, in dump
    _LokyPickler(file, reducers=reducers, protocol=protocol).dump(obj)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/cloudpickle/cloudpickle.py", line 267, in dump
    return Pickler.dump(self, obj)
  File "/usr/lib/python3.6/pickle.py", line 409, in dump
    self.save(obj)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 852, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 781, in save_list
    self._batch_appends(obj)
  File "/usr/lib/python3.6/pickle.py", line 808, in _batch_appends
    save(tmp[0])
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 736, in save_tuple
    save(element)
  File "/usr/lib/python3.6/pickle.py", line 496, in save
    rv = reduce(self.proto)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/loky/cloudpickle_wrapper.py", line 20, in __reduce__
    _pickled_object = dumps(self._obj)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/cloudpickle/cloudpickle.py", line 952, in dumps
    cp.dump(obj)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/cloudpickle/cloudpickle.py", line 267, in dump
    return Pickler.dump(self, obj)
  File "/usr/lib/python3.6/pickle.py", line 409, in dump
    self.save(obj)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/cloudpickle/cloudpickle.py", line 718, in save_instancemethod
    self.save_reduce(types.MethodType, (obj.__func__, obj.__self__), obj=obj)
  File "/usr/lib/python3.6/pickle.py", line 610, in save_reduce
    save(args)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 736, in save_tuple
    save(element)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 781, in save_list
    self._batch_appends(obj)
  File "/usr/lib/python3.6/pickle.py", line 808, in _batch_appends
    save(tmp[0])
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 736, in save_tuple
    save(element)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 736, in save_tuple
    save(element)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 736, in save_tuple
    save(element)
  File "/usr/lib/python3.6/pickle.py", line 496, in save
    rv = reduce(self.proto)
TypeError: can't pickle weakref objects
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "./freqtrade/main.py", line 60, in <module>
Exception in thread QueueFeederThread:
Traceback (most recent call last):
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/loky/backend/queues.py", line 150, in _feed
    obj_ = dumps(obj, reducers=reducers)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/loky/backend/reduction.py", line 243, in dumps
    dump(obj, buf, reducers=reducers, protocol=protocol)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/loky/backend/reduction.py", line 236, in dump
    _LokyPickler(file, reducers=reducers, protocol=protocol).dump(obj)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/cloudpickle/cloudpickle.py", line 267, in dump
    return Pickler.dump(self, obj)
  File "/usr/lib/python3.6/pickle.py", line 409, in dump
    self.save(obj)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 852, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 781, in save_list
    self._batch_appends(obj)
  File "/usr/lib/python3.6/pickle.py", line 808, in _batch_appends
    save(tmp[0])
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 736, in save_tuple
    save(element)
  File "/usr/lib/python3.6/pickle.py", line 496, in save
    rv = reduce(self.proto)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/loky/cloudpickle_wrapper.py", line 20, in __reduce__
    _pickled_object = dumps(self._obj)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/cloudpickle/cloudpickle.py", line 952, in dumps
    cp.dump(obj)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/cloudpickle/cloudpickle.py", line 267, in dump
    return Pickler.dump(self, obj)
  File "/usr/lib/python3.6/pickle.py", line 409, in dump
    self.save(obj)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/cloudpickle/cloudpickle.py", line 718, in save_instancemethod
    self.save_reduce(types.MethodType, (obj.__func__, obj.__self__), obj=obj)
  File "/usr/lib/python3.6/pickle.py", line 610, in save_reduce
    save(args)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 736, in save_tuple
    save(element)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 781, in save_list
    self._batch_appends(obj)
  File "/usr/lib/python3.6/pickle.py", line 808, in _batch_appends
    save(tmp[0])
  File "/usr/lib/python3.6/pickle.py", line 521, in save
    self.save_reduce(obj=obj, *rv)
  File "/usr/lib/python3.6/pickle.py", line 634, in save_reduce
    save(state)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 736, in save_tuple
    save(element)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 821, in save_dict
    self._batch_setitems(obj.items())
  File "/usr/lib/python3.6/pickle.py", line 847, in _batch_setitems
    save(v)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 736, in save_tuple
    save(element)
  File "/usr/lib/python3.6/pickle.py", line 476, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python3.6/pickle.py", line 736, in save_tuple
    save(element)
  File "/usr/lib/python3.6/pickle.py", line 496, in save
    rv = reduce(self.proto)
TypeError: can't pickle weakref objects

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/threading.py", line 916, in _bootstrap_inner
    self.run()
  File "/usr/lib/python3.6/threading.py", line 864, in run
    self._target(*self._args, **self._kwargs)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/loky/backend/queues.py", line 175, in _feed
    onerror(e, obj)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/loky/process_executor.py", line 310, in _on_queue_feeder_error
    self.thread_wakeup.wakeup()
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/externals/loky/process_executor.py", line 155, in wakeup
    self._writer.send_bytes(b"")
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 183, in send_bytes
    self._check_closed()
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 136, in _check_closed
    raise OSError("handle is closed")
OSError: handle is closed

    main(sys.argv[1:])
  File "./freqtrade/main.py", line 34, in main
    args.func(args)
  File "/root/freqtrade/freqtrade/optimize/hyperopt.py", line 354, in start
    hyperopt.start()
  File "/root/freqtrade/freqtrade/optimize/hyperopt.py", line 308, in start
    f_val = self.run_optimizer_parallel(parallel, asked)
  File "/root/freqtrade/freqtrade/optimize/hyperopt.py", line 271, in run_optimizer_parallel
    wrap_non_picklable_objects(self.generate_optimizer))(v) for v in asked)
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/parallel.py", line 934, in __call__
    self.retrieve()
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/parallel.py", line 833, in retrieve
    self._output.extend(job.get(timeout=self.timeout))
  File "/root/freqtrade/.env/lib/python3.6/site-packages/joblib/_parallel_backends.py", line 521, in wrap_future_result
    return future.result(timeout=timeout)
  File "/usr/lib/python3.6/concurrent/futures/_base.py", line 432, in result
    return self.__get_result()
  File "/usr/lib/python3.6/concurrent/futures/_base.py", line 384, in __get_result
    raise self._exception
_pickle.PicklingError: Could not pickle the task to send it to the workers.

  ```

**Comment by hroff-1902** [comment]
What version of joblib you run the bot with?

**Comment by xmatthias** [comment]
> TypeError: can't pickle weakref objects

the key is the above... do you have anything "special" in your hyperopts file? (odd imports, instanciating something special, ...)?

**Comment by coder194** [comment]
> hroff-1902

I am running joblib 0.13.2      

> xmatthias

 I downloaded and the original hyperopt.py file from the github repository and used this for running the optimization. Thankfully the optimization now works, however, I get the following warning message when using 2 CPUs:

```
/root/freqtrade/.env/lib/python3.6/site-packages/sklearn/externals/joblib/__init__.py:15: DeprecationWarning: sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. Please import this functionality directly from joblib, which can be installed with: pip install joblib. If this warning is raised when loading pickled models, you may need to re-serialize those models with scikit-learn 0.21+.
  warnings.warn(msg, category=DeprecationWarning)
/root/freqtrade/.env/lib/python3.6/site-packages/sklearn/externals/joblib/__init__.py:15: DeprecationWarning: sklearn.externals.joblib is deprecated in 0.21 and will be removed in 0.23. Please import this functionality directly from joblib, which can be installed with: pip install joblib. If this warning is raised when loading pickled models, you may need to re-serialize those models with scikit-learn 0.21+.
  warnings.warn(msg, category=DeprecationWarning)
```

May you please confirm whether I can still proceed with successfully running the optimization despite these warnings?

**Comment by xmatthias** [comment]
this is just a DeprecationWarning thrown by sklearn, they're telling us that this won't work sometime in the future - nothing to worry about for now.

**Comment by hroff-1902** [comment]
> Thankfully the optimization now works

What has changed?

**Comment by coder194** [comment]
hroff-1902: I am now able to successfully run the optimization and get an output of the optimized parameters for my strategy. 

xmatthias: Sounds good - will keep an eye out for this. 

Thank you again for the help!

**Comment by hroff-1902** [comment]
> hroff-1902: I am now able to successfully run the optimization and get an output of the optimized parameters for my strategy.

What has changed that previously you was not able to run the bot and now you can?

**Comment by coder194** [comment]
Hello hroff-1902: before when I allocated more than one CPU to the optimization process, the code would break and I was not able to get a list of the optimized parameters. However, using the suggestions mentioned above, I am able to now run the optimization successfully and get a list of the optimized values. What I changed was the content within the hyperopts.py file which I just re-downloaded from the github link and replaced it (I made some edits to it initially which may have messed it up).
